flybywiresim/a32nx
# fix(mcdu): take off page THR RED ACC and ENG OUT ACC calculations and restrictions

Fix for issue #6271 Engine out acceleration altitude is not using airfield elevation in the calculation on the MCDU perf take off page.
It also adds restrictions to the THR RED ACC values that can be entered.

## Summary of Changes
 origin.altitudeinFP no longer returns the elevation of the orgin airfield.  It returns zero. This results in the minimum altitude restriction for THR RED ACC and ENG OUT ACC to not take account of airfield elevation.  Additionally no restrictions were imposed on the valid values for THR RED and ACC. This allowed ACC to be set lower than THR RED. This PR sets the minimum performance altitudes to FROM airfield elevaion + 400 and imposses that ACC => THR RED. Only the take off page has been modified in this PR. The go around also has the same issues, I suggest these should be addressed when the GA track is implemented.
 
## Screenshots (if necessary)
![image](https://user-images.githubusercontent.com/31592841/147797674-8ba08fef-bf3e-4058-8b3d-b7f2257e5905.png)


## References
FCOM. Verified in discussions with type rated A320 pilots. 

## Additional context
Other parts of the code make use of xxxxx.altitudeInFP these should be checked as unlikely they work as intended.

john P Maguire#9027
Discord username (if different from GitHub):

## Testing instructions
Load in on the stand and on the runway. (take off page performance altitudes should not be editable after take off).

1. In the EFB check / set an offset value for THR RED ACC and ENG OUT ACC. Defaults to 1500. Change this during the test to check that the entered offset is used.

2. Check that without a FROM/TO in INIT A the performance attitudes display ----/---- and ---- 

3. On entering a FROM/TO on INIT A the THR RED ACC and ENG OUT ACC altitudes are calculated using the EFB offset + the airfield elevation of the FROM airfield.

4. Enter a different value for each of THR RED ACC and ENG OUT ACC (a pilot entered value). Check that four digits are required eg eng out acc 900, 0900 must be entered.  Check the text size is changed to large. Use CLR to clear the value and check it reverts back to the calculated altitude and small text. 

5. Check that a lower altitude cannot be entered for ACC than THR RED or a higher altitude cannot be entered for THR RED than ACC. These values should be able to be entered as a pair or individually. 

6. Change the FROM/TO in INIT A check that the new FROM airfield altitude is used and an updated THR RED ACC and ENG OUT ACC is correctly calculated for the new FROM airfield.

7. Enter a pilot value for the THR RED ACC and ENG OUT ACC. Change the FROM/TO airports on INIT A and check that the pilot entered value is not changed until the CLR is used to reset the calculation.  This should revert to small text and the correct altitude (airfield elevation + EFB offset).

8. Check that an altitude lower than the FROM airfield elevation + 400 cannot be entered. That a value higher than 45 000 cannot be entered in any of the fields THR RED ACC or ENG OUT ACC.

9. Check for no regressions.


## How to download the PR for QA

Every new commit to this PR will cause a new A32NX artifact to be created, built, and uploaded.

1. Make sure you are signed in to GitHub
1. Click on the **Checks** tab on the PR
1. On the left side, click on the bottom **PR** tab
1. Click on the **A32NX** download link at the bottom of the page
